Libs/MRML/Core/vtkMRMLStorageNode.h

```diff
@@ -26,6 +27,76 @@ class vtkStringArray;
 
 // STD includes
 #include <vector>
+#include <algorithm>
+
+///
+/// A class for recording warnings and errors associated with this
+/// vtkMRMLStorableNode.  This cannot be hidden within the
+/// vtkMRMLStorableNode class scope because the Python wrapping code
+/// vtkMRMLStorageNodePython.cxx cannot wrap the GetVTKMessageLog
+/// method.
+class vtkMessageLog
+{
+public:
+  using event_type = unsigned long;
+  ///
+  /// A class that describes a single message
+  struct Message
+  {
+    Message(event_type eventId, const std::string &messageText)
+      : EventId(eventId), MessageText(messageText) {}
+    ///
+    /// Function object for checking whether a Message has a specified
+    /// EventID.
+    struct EqualEventId
+    {
+      EqualEventId(event_type eventId) : EventId(eventId) {}
+      bool operator()(const Message &message) const
+        { return message.EventId == this->EventId; }
+
+      event_type EventId;
+    };
+
+    event_type EventId;
+    std::string MessageText;
+  };
+
+  ///
+  /// The kind of integer that is used for indexing the message vector
+  using size_type = typename std::vector<Message>::size_type;
```

Comment:
On my system, `std::vector<Message>::size_type` is `unsigned long`.  I will use `int`, but if it should be `unsigned` or `unsigned long`, please let me know.

Reply:
Using the `vtkVariantArray` or other standard vtk container would sidestep this issue.

Reply:
The more I think about it, two `vtkStringArray`s like `ErrorMessages` and `WarningMessages` instance variables might actually be all we need.  This would get rid of the helper class, and also provide the python wrapping and observability.  What do you think of this simpler solution?

Reply:
Pros: `vtkStringArray` provides all sorts of methods that have proved useful, at least in other circumstances.  Its integer type for indexing is already fixed independent of the system on which it is compiled (unlike the case for `std::vector<...>`), so we will not have to worry about issues relating to exposing `int` when the underlying `std::vector<...>` instead supports its own `size_type`.

Cons: having exactly two instances of `vtkStringArray` supports exactly two message types, errors and warnings; in this storage node context or some future context will we want to support a larger or flexible subset of the message types from `VTK/Common/Core/vtkCommand.h`?

Proposal: we leave this as is for now (i.e., using a `std::vector<typename vtkMessages::Message>` object rather than a pair of `vtkStringArray` objects) and we revisit once we have implemented the `Slicer/Base/QTGUI/qSlicerSaveDataDialog.cxx` side of things.  It is plausible that the needs in `qSlicerSaveDataDialog.cxx` will advise this situation.

Reply:
Sure, it's not technically wrong the way it is now, and yes, it's good to prototype the whole thing to get a feel for how all the pieces work.

I'll still push for re-using existing VTK concepts wherever possible, not just for the python and event benefits, but also it's more maintainable and readable in a large code base because people know the naming scheme, trust the implementation isn't buggy, know it will compile across platforms, etc.  The slicer policy is to follow the coding conventions of the parent class, and VTK is good at this.  Here you might use an std map of event ids to vtkStringArray for example, so there's only one instance variable; this pattern is used a lot in VTK code.

Reply:
The `std::map` approach, etc. is now discussed in another thread.  I am marking this thread as resolved.